src/lib.rs

```diff
@@ -1,126 +1,82 @@
 //! and services that use them. The interface is based on the `std::futures` API.
-//!
-//! ## Example
-//! ```rust,no_run
-//! use futures::{
-//!     future::{self, BoxFuture, FutureExt},
-//! };
-//! use http_service::{HttpService, Response};
-//! use std::net::{IpAddr, Ipv4Addr, SocketAddr};
-//!
-//! struct Server {
-//!     message: Vec<u8>,
-//! }
-//!
-//! impl Server {
-//!     fn create(message: Vec<u8>) -> Server {
-//!         Server {
-//!             message,
-//!         }
-//!     }
-//!
-//!     pub fn run(s: Server) {
-//!         let a = SocketAddr::new(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)), 8080);
-//!         http_service_hyper::run(s, a);
-//!     }
-//! }
-//!
-//! impl HttpService for Server {
-//!     type Connection = ();
-//!     type ConnectionFuture = future::Ready<Result<(), std::io::Error>>;
-//!     type ResponseFuture = BoxFuture<'static, Result<http_service::Response, std::io::Error>>;
-//!
-//!     fn connect(&self) -> Self::ConnectionFuture {
-//!         future::ok(())
-//!     }
-//!
-//!     fn respond(&self, _conn: &mut (), _req: http_service::Request) -> Self::ResponseFuture {
-//!         let message = self.message.clone();
-//!         async move { Ok(Response::new(http_service::Body::from(message))) }.boxed()
-//!     }
-//! }
-//!
-//! fn main() {
-//!     let s = Server::create(String::from("Hello, World").into_bytes());
-//!     Server::run(s);
-//! }
-//! ```
-
-#![forbid(future_incompatible, rust_2018_idioms)]
-#![deny(missing_debug_implementations, nonstandard_style)]
-#![warn(missing_docs, missing_doc_code_examples)]
-#![cfg_attr(any(feature = "nightly", test), feature(external_doc))]
-#![cfg_attr(feature = "nightly", doc(include = "../README.md"))]
-#![cfg_attr(test, deny(warnings))]
-
-use bytes::Bytes;
-use futures::{
-    future,
-    prelude::*,
-    stream::{self, BoxStream},
-    task::{Context, Poll},
-};
+
+#![warn(missing_debug_implementations, rust_2018_idioms)]
+#![allow(clippy::mutex_atomic, clippy::module_inception)]
+#![doc(test(attr(deny(rust_2018_idioms, warnings))))]
+#![doc(test(attr(allow(unused_extern_crates, unused_variables))))]
+
+use async_std::io::{self, prelude::*};
+use async_std::task::{Context, Poll};
+
+use futures::future::TryFuture;
 
 use std::fmt;
 use std::pin::Pin;
 
-#[cfg(test)]
-#[doc(include = "../README.md")]
-const _README: () = ();
-
-/// The raw body of an http request or response.
-///
-/// A body is a stream of `Bytes` values, which are shared handles to byte buffers.
-/// Both `Body` and `Bytes` values can be easily created from standard owned byte buffer types
-/// like `Vec<u8>` or `String`, using the `From` trait.
-pub struct Body {
-    stream: BoxStream<'static, Result<Bytes, std::io::Error>>,
+pin_project_lite::pin_project! {
+    /// The raw body of an http request or response.
+    pub struct Body {
+        #[pin]
+        reader: Box<dyn BufRead + Unpin + Send + 'static>,
```

Comment:
Boxes are always `Unpin` so we can just `Pin::new(&mut self.reader)`. We don't even need `Unpin` bound for inner type!

Reply:
`Box<dyn BufRead>` doesn't implement `BufRead` though, this should be `Pin<Box<dyn BufRead>>` to avoid the `Unpin` bound.

Reply:
Oh that's a great, thanks for pointing that out!